README.md
# STREAM ONLINE VIEWER

**STREAM ONLINE VIEWER** is an application to allow multiple clients to display a live stream and visualize it on a web client via browser.

### Definition
-   Multiple clients can watch the stream on one web server (multiple web clients connected).
- The server should be always listening to the stream for messages and discard this messages if there are no clients to send the message to.
- Each client should start receiving images from the moment he connects onward (do not send old images that the server received before the client connected).
- When a user opens the web page, it should automatically start receiving the latest data from the server. The web page should display to the user the image and metadata from the stream.

### Architecture & Technology stack
Angular + Python + Flask

- Angular is JavaScript (well, actually written in Typescript) framework for building Single Page Application.
- Python is server side programming language and very powerful to build complex IoT, Image Processing and other types of system.
- Flask is a REST API framework for Python.

For communication between server and clients, Flask-SocketIO(https://flask-socketio.readthedocs.io/en/latest/) has been used. It gives low latency bi-directional communications between the clients and the server.

For the front end design, Materialize(https://materializecss.com/) was used. It is a modern responsive front-end framework based on Material Design (https://material.io/), developed by Google.


## Installation
To install the STREAM ONLINE VIEWER clone the repository and set up the anaconda virtual environment based on the requirements.txt file in combination with pip and virtualenv.

```bash
git clone https://github.com/lhdamiani/interview_stream_online_viewer
cd interview_stream_online_viewer
mkvirtualenv -p /usr/local/anaconda3/bin/python interview
pip install -r requirements.txt 
```
> **Note**: if mkvirtualenv is not found, make sure to set the environment variable VIRTUALENVWRAPPER_PYTHONto pointing to `export VIRTUALENVWRAPPER_PYTHON=/usr/local/anaconda3/bin/python`


### Dependencies
In order to use this repository, you will have to install the following conda packages:

- matplotlib
- numpy
- bsread

You can do this (on Linux) by running:
```bash
conda install -c paulscherrerinstitute bsread numpy matplotlib
```

When installing PSI libraries you have to specify that you want to install them from the PSI Anaconda repository (the "-c paulscherrerinstitute" flag in the command above).

> **Note**: if conda is not found, make sure to add to the path `export PATH=/Users/lhdamiani/miniconda3/bin:$PATH`, don't forget to change the username.

## Run
-------------------------
To run the Flask-Socketio web server from the ROOT folder:

```bash
export PYTHONPATH=$(pwd):${PYTHONPATH}
python stream_online_viewer/start_server.py -H <HOST> -P <PORT> -S <SOURCE_STREAM_PORT> -O <SOURCE_STREAM_HOST>
```
> **Note**: Default values for `<HOST>`, `<PORT>`, `<SOURCE_STREAM_PORT>` and `<SOURCE_STREAM_HOST>` are 127.0.0.1, 5000, 8888, and 127.0.0.1 respectively.

To run the data generator:
```bash
export PYTHONPATH=$(pwd):${PYTHONPATH}
python stream_online_viewer/start_stream.py
```
> **Note**: Default values for the data generated is 8888

To access the client viewer go to 127.0.0.1:5000 on your browser
> **Note**: Firefox is recommended.

and proceed with the login

**Username: admin**

**Password: password**

> **Note**: For the purpose of this task, not much effort has been invested into the security method. The method adopted simply serves as a reminder that such web app needs security.

The client works in any screen-width - it is ready for mobile usage and small screens.

### Data stream generator

This is the protocol we use to transfer beam synchronous data at SwissFEL: [https://github.com/paulscherrerinstitute/bsread_python](https://github.com/paulscherrerinstitute/bsread_python)

> **Note:** You will not be able to access the real BSread stream from outside of PSI. This repository has a stream generator you can use to simulate camera images and metadata you will later display to the clients. 


The generator is located in **stream_online_viewer/start_stream.py** and can be run from the ROOT of this repository:
```
export PYTHONPATH=$(pwd):${PYTHONPATH}
python stream_online_viewer/start_stream.py
```


### Possible improvements and drawbacks

|What?|Current implementation|How to improve?|
|----------------|-------------------------------|-----------------------------|
|Performance| Due to time and purpose of this task, the focus was to implement a minimally running example with the desired functionality. Flask-SocketIO was used to communication between server and clients. Analysis and benchmarks analysing the performance of Flask-SocketIO have been performed by others and provide a satisfactory results for this application. Such analysis can be found in http://drewww.github.io/socket.io-benchmarking/ and http://blog.mixu.net/2011/11/22/performance-benchmarking-socket-io-0-8-7-0-7-11-and-0-6-17-and-nodes-native-tcp/| Deeper analysis about possible tools can be further analyzed once the time is not a constraint. |
|Security|Just a simple login/password to block the visualization of the web app contet.| Implementation of security using Flask-Security(https://pythonhosted.org/Flask-Security/) or Flask-Oauth(https://pythonhosted.org/Flask-OAuth/)|
|Image loading|For this tasks purpose, the web server receives the message from the data stream generator and saves it to a local file image which is refreshed on the client side's browser. Probably with a high rate of messages it could be a problem to show them with a high frequency. |Transmission of the image data by some encoding protocol which would transfer the data it on a more efficient way without the cost of writing to a disk and re-loading on the client's side. Encoding/Decoding protocols as B64 could be of use https://www.base64-image.de/.|
|Mobile access and different browsers| Current implementation was focused on regular computer 'width' screens but implemented with angular and materialize which allows small screens to access the content in a responsive way. But, due to the time constraint and purpose of this task, no extensive tests have been made in order to assure that all functionalities are flawless. Firefox browser was used for development.| Perform tests and evaluate the usability and functionalities on small-screens devices and different browsers.|


### Acknowledgement

This application was loosely developed based on this repository: https://github.com/miguelgrinberg/Flask-SocketIO


### Changelog

See file `CHANGES.rst`.

## Contact

For comments, contact:
```
leonardo.hax@psi.ch
```

CHANGES.rst
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## [0.0.3] - 2018-10-08
### Changed
- Explanation of the architecture and analysis of drawbacks/performance into README file.

## [0.0.2] - 2018-10-07
### Changed
- Parametrization of the server.
- Adjustment of the plot.
- Bugfix: img id source for loading correctly the image.

### Added
- Implementation of the logout restful resource api.
- Insertion of the collapsible information.
- Showing meaningful information on cards when receiving stream.
- Connection with the image generated
- New 404 page.
- Connections with the stream data generator.

## [0.0.1] - 2018-10-04

### Added
- This CHANGELOG file.
- README.md file.
- requirements.txt file for easy installation.
- web_server.py containing the source code of the flask-socketio web server.
- templates folder including login, index and 404 html files.
- images folder containing the nav-sidebar images office.png and pic.png.
- css folder containing the css styling files.
- js folder containing the javascript files.
- lib folder containing angular, bootstrap, flot and jquery files. 

[Unreleased]: https://github.com/paulscherrerinstitute/interview_stream_online_viewer/compare/master...lhdamiani:master
[0.0.3]: https://github.com/lhdamiani/interview_stream_online_viewer/tree/v0.0.3
[0.0.2]: https://github.com/lhdamiani/interview_stream_online_viewer/tree/v0.0.2
[0.0.1]: https://github.com/lhdamiani/interview_stream_online_viewer/tree/v0.0.1




tests/test_data.py
import unittest

from numpy.testing import assert_array_equal

from stream_online_viewer.data import generate_image, get_stream_message_data, generate_metadata


class TestUtils(unittest.TestCase):

    def test_image(self):
        expected_size_x = 1000
        expected_size_y = 2000

        image = generate_image(size_x=expected_size_x, size_y=expected_size_y)

        # Slowest changing dimension first.
        self.assertEqual(image.shape[0], expected_size_y)
        self.assertEqual(image.shape[1], expected_size_x)

    def test_metadata(self):
        beam_energy_base = 100
        repetition_rate = 112

        metadata = generate_metadata(beam_energy_base=beam_energy_base, repetition_rate=repetition_rate)

        self.assertIn("beam_energy", metadata)
        # The beam energy fluctuates from the base value to + 10% max.
        self.assertTrue(beam_energy_base <= metadata["beam_energy"] <= beam_energy_base*1.1)

        self.assertIn("repetition_rate", metadata)
        self.assertEqual(metadata["repetition_rate"], repetition_rate)

    def test_stream_simulated_data(self):
        expected_size_x = 1024
        expected_size_y = 512
        beam_energy_base = 50
        repetition_rate = 25

        image = generate_image(size_x=expected_size_x, size_y=expected_size_y)
        metadata = generate_metadata(beam_energy_base=beam_energy_base, repetition_rate=repetition_rate)

        stream_data = get_stream_message_data(image=image, metadata=metadata)

        assert_array_equal(stream_data["image"], image)

        self.assertEqual(stream_data["image_size_x"], expected_size_x)
        self.assertEqual(stream_data["image_size_y"], expected_size_y)

        self.assertEqual(stream_data["image_profile_x"].shape, (expected_size_x,))
        self.assertEqual(stream_data["image_profile_y"].shape, (expected_size_y,))

        self.assertTrue(beam_energy_base <= metadata["beam_energy"] <= beam_energy_base * 1.1)
        self.assertEqual(metadata["repetition_rate"], repetition_rate)


tests/test_stream.py
import unittest

from multiprocessing import Process

from bsread import source

from stream_online_viewer.stream import generate_stream


class TestGenerateStream(unittest.TestCase):

    def setUp(self):
        self.stream_output_port = 8888
        self.n_images = 10

        self.stream_generator_process = Process(target=generate_stream,
                                                args=(self.stream_output_port, self.n_images))
        self.stream_generator_process.start()

    def tearDown(self):
        self.stream_generator_process.terminate()

    def test_generate_stream(self):

        message = None

        # You always need to specify the host parameter, otherwise bsread will try to access PSI servers.
        with source(host="localhost", port=self.stream_output_port, receive_timeout=1000) as input_stream:

            n_received = 1

            while n_received < self.n_images:
                message = input_stream.receive()

                # In case of receive timeout (1000 ms in this example), the received data is None.
                if message is None:
                    continue

                n_received += 1

        self.assertIsNotNone(message)

        image = message.data.data["image"].value
        image_size_x = message.data.data["image_size_x"].value
        image_size_y = message.data.data["image_size_y"].value

        self.assertEqual(image.shape[0], image_size_y)
        self.assertEqual(image.shape[1], image_size_x)

        image_profile_x = message.data.data["image_profile_x"].value
        self.assertEqual(image_profile_x.shape[0], image_size_x)

        image_profile_y = message.data.data["image_profile_y"].value
        self.assertEqual(image_profile_y.shape[0], image_size_y)

        beam_energy = message.data.data["beam_energy"].value
        self.assertGreater(beam_energy, 0)

        repetition_rate = message.data.data["repetition_rate"].value
        self.assertGreater(repetition_rate, 0)


stream_online_viewer/data.py
from random import random

import numpy
from matplotlib import pyplot


def get_stream_message_data(image, metadata):
    """
    Create stream data payload from provided image and metadata.
    :param image: Image to prepare the message data for.
    :param metadata: Metadata to add to the message.
    :return: Dictionary with stream message payload.
    """

    stream_data = {
        "image": image,
        "image_size_x": image.shape[1],
        "image_size_y": image.shape[0],
        "image_profile_x": image.sum(0),
        "image_profile_y": image.sum(1)
    }

    stream_data.update(metadata)

    return stream_data


def generate_image(size_x=640, size_y=480, beam_size_x=10, beam_size_y=8, noise=0.1):
    """
    Generate a 2D numpy array with a simulated beam image.
    :param size_x: Width of the generated image.
    :param size_y: Height of the generated image.
    :param beam_size_x: Width of the simulated beam in the image.
    :param beam_size_y: Height of the simulated beam in the image.
    :param noise: Noise level to introduce to the image.
    :return: Numpy array with the image.
    """

    beam_x = numpy.linspace(-beam_size_x + numpy.random.rand(),
                            beam_size_x + numpy.random.rand(),
                            size_y)

    beam_y = numpy.linspace(-beam_size_y + numpy.random.rand(),
                            beam_size_y + numpy.random.rand(),
                            size_x)

    x, y = numpy.meshgrid(beam_y, beam_x)

    image = numpy.exp(-(x ** 2 + y ** 2))

    image += numpy.random.random((size_y, size_x)) * noise
    # display_image(image)
    return image


def generate_metadata(beam_energy_base=5, repetition_rate=100):
    """
    Generate metadata to be send over with every image.
    :return: Dictionary of metadata.
    """
    return {
        "beam_energy": beam_energy_base + ((beam_energy_base*0.1) * random()),
        "repetition_rate": repetition_rate
    }


def display_image(image):
    """
    Debug method to display image on screen.
    :param image: Image to display.
    """
    pyplot.imshow(image)
    pyplot.show()


stream_online_viewer/static/js/application.js
$(document).ready(function(){

    $('input.autocomplete').autocomplete({
      data: {
        "Apple": null,
        "Microsoft": null,
        "Google": 'https://placehold.it/250x250'
      },
    });

    $('.collapsible').collapsible();
    $('.sidenav').sidenav();
    $('.parallax').parallax();
    $(".dropdown-trigger").dropdown();
    $('.materialboxed').materialbox();


    var data = [];
    var dataset;
    var updateInterval = 1000;
    var now = new Date().getTime();
    var timer;
    // resizes the window range of the interesting plot with beam energy
    $(window).bind('resize', function() {
    clearTimeout(timer);
    timer = setTimeout(function(){ $(window).resize(); }, 250);
    });

    // Options for the interesting plot with beam energy
    var options = {
        series: {
            lines: {
                show: true,
                lineWidth: 1.2,
                fill: true
            }
        },
        xaxis: {
            mode: "time",
            tickSize: [2, "second"],
            tickFormatter: function (v, axis) {
                var date = new Date(v);
    
                if (date.getSeconds() % 10 == 0) {
                    var hours = date.getHours() < 10 ? "0" + date.getHours() : date.getHours();
                    var minutes = date.getMinutes() < 10 ? "0" + date.getMinutes() : date.getMinutes();
                    var seconds = date.getSeconds() < 10 ? "0" + date.getSeconds() : date.getSeconds();
    
                    return hours + ":" + minutes + ":" + seconds;
                } else {
                    return "";
                }
            },
            axisLabel: "Time",
            axisLabelUseCanvas: true,
            axisLabelFontSizePixels: 12,
            axisLabelFontFamily: 'Verdana, Arial',
            axisLabelPadding: 10
        },
        yaxis: {
            min: 5,
            max: 6,        
            tickSize: 0.25,
            tickFormatter: function (v, axis) {
                if (v % .25 == 0) {
                    return v;
                } else {
                    return "";
                }
            },
            axisLabel: "Beam energy",
            axisLabelUseCanvas: true,
            axisLabelFontSizePixels: 12,
            axisLabelFontFamily: 'Verdana, Arial',
            axisLabelPadding: 6
        },
        legend: {        
            labelBoxBorderColor: "#fff",
            show: false
        },
        grid: {                
            backgroundColor: "#fff",
            tickColor: "#008040",
            hoverable: true
        }
    };

    //connect to the socket server.
    var socket = io.connect('http://' + document.domain + ':' + location.port + '/test');
    var messages_received = [];

    // Update the interesting plot with beam energy
    function update(){
        $.plot($("#flot-placeholder1"), dataset, options);
        
    }

    //receive details from server
    socket.on('newmessage', function(msg) {
        messages_received.push(msg.number_of_received_messages);
        // Collapsible image details html
        imgDetails = "Image size x:"+ msg.image_size_x.toString() +"(originally) <br> Image size y:" + msg.image_size_y.toString() +"(originally)"
        $('#ImgDetails').html(imgDetails);
        // Collapsible connection details html
        connectionDetails = "Total bytes received: "+msg.total_bytes_received.toString()+ "<br>Numbers of received messages: " +msg.number_of_received_messages.toString()
        $('#ConnectionDetails').html(connectionDetails);
        // Collapsible metadata details html
        metadataDetails = "Beam energy:"+ msg.beam_energy.toString() +"<br> Repetition rate:" + msg.repetition_rate.toString()
        $('#MetadataDetails').html(metadataDetails);
        // Src for the image from server
        valueTime = new Date().getTime()
        $("#imgFromServer").attr("src","/static/images/stream.png?t=" + valueTime);
        // Updates the data for the interesting plot with beam energy
        if (messages_received.length >= 15){
            data.shift();
        } 
        data.push([now += updateInterval, msg.beam_energy])
        dataset = [{ label: "Info", data: data, color: "#00FF00" }];
        update()
        setTimeout(update, updateInterval);
    });

});



(function () {
    'use strict';
  
    angular.module('StreamApp', [])
  
    .controller('StreamController', ['$scope', '$log', '$http',
      function($scope, $log, $http) {
      }
    ]);
  
  }());

'use strict';

angular.module('StreamApp', ['angularFlaskServices'])
	.config(['$routeProvider', '$locationProvider',
		function($routeProvider, $locationProvider) {
		$routeProvider
		.when('/', {
			templateUrl: 'static/partials/landing.html',
			controller: IndexController
		})
		.when('/about', {
			templateUrl: '',
			controller: AboutController
		})
		.when('/logout', {
			templateUrl: '',
			controller: LogoutController
		})
		.otherwise({
			redirectTo: '/'
		})
		;

		$locationProvider.html5Mode(true);	

	}]);

	$(function() {
		setTimeout(function(){
			$(".message_flash").hide('blind', {}, 500)
		},1000);
	})

stream_online_viewer/start_server.py

# Start with a basic flask app webpage.
from flask_socketio import SocketIO, emit
from flask_restful import Resource, Api
from flask_material import Material  
from flask_cors import CORS
from flask import Flask, flash, jsonify, render_template, url_for, copy_current_request_context, request, make_response, session, redirect, abort, _request_ctx_stack
from threading import Thread, Event
import optparse
from bsread import source
from matplotlib import pyplot, image

__author__ = 'hax_damiani'

app = Flask(__name__)
Material(app)
CORS(app)

api = Api(app)
app.config['SECRET_KEY'] = 'HaxDamiani!'
app.config['DEBUG'] = True

#turn the flask app into a socketio app
socketio = SocketIO(app)

# Server thread
thread = Thread()
thread_stop_event = Event()


class ClientThread(Thread):
    def __init__(self, port, n_img, stream_host, ):
        super(ClientThread, self).__init__()
        self._delay = 1.5
        self._stream_output_port = port
        self._n_images = n_img
        self._stream_host = stream_host

    def receive_stream(self):
        """
        Function that receives the stream and send the signal for the clients using socketio.
        """
        message = None
        # You always need to specify the host parameter, otherwise bsread will try to access PSI servers.
        with source(host=self._stream_host, port=self._stream_output_port, receive_timeout=1000) as input_stream:

            n_received = 0
            # Detects how many messages are expected
            if self._n_images == -1:
                while True:
                    message = input_stream.receive()
                    # In case of receive timeout (1000 ms in this example), the received data is None.
                    if message is None:
                        continue
                    else:
                        # Creates the image and saves to the file that is shown to the client
                        pyplot.imshow(message.data.data['image'].value)
                        pyplot.savefig('./stream_online_viewer/static/images/stream.png')
                        # Increases the number of received messages
                        n_received += 1
                        # Generates the data containing meaningful information to the client
                        data = {'number_of_received_messages':  n_received, 
                                'data': n_received,
                                'messages_received': float(message.statistics.messages_received),
                                'total_bytes_received': float(message.statistics.total_bytes_received),
                                'repetition_rate': float(message.data.data['repetition_rate'].value),
                                'beam_energy': float(message.data.data['beam_energy'].value),
                                'image_size_y': float(message.data.data['image_size_y'].value),
                                'image_size_x': float(message.data.data['image_size_x'].value)
                                }
                        # emits the signal with the data
                        socketio.emit('newmessage', data, namespace='/test')
            else:
                for _ in range(self._n_images):
                    message = input_stream.receive()
                    # In case of receive timeout (1000 ms in this example), the received data is None.
                    if message is None:
                        continue
                    else:
                        # Creates the image and saves to the file that is shown to the client
                        pyplot.imshow(message.data.data['image'].value)
                        pyplot.savefig('./stream_online_viewer/static/images/stream.png')
                        # Increases the number of received messages
                        n_received += 1
                        # Generates the data containing meaningful information to the client
                        data = {'number_of_received_messages':  n_received, 
                                'data': n_received,
                                'messages_received': float(message.statistics.messages_received),
                                'total_bytes_received': float(message.statistics.total_bytes_received),
                                'repetition_rate': float(message.data.data['repetition_rate'].value),
                                'beam_energy': float(message.data.data['beam_energy'].value),
                                'image_size_y': float(message.data.data['image_size_y'].value),
                                'image_size_x': float(message.data.data['image_size_x'].value)
                                }
                        # emits the signal with the data
                        socketio.emit('newmessage', data, namespace='/test')

    def run(self):
        self.receive_stream()
        

# Handles the 404 error and renders the 404 template.
@app.errorhandler(404)
def page_not_found(e):
    return render_template('404.html'), 404

# Detects a new client connected and start the thread.
@socketio.on('connect', namespace='/test')
def test_connect():
    # need visibility of the global thread object
    global thread
    print('Client connected...')
    

    #Start the random number generator thread only if the thread has not been started before.
    if not thread.isAlive():
        print("Starting Thread")
        thread = ClientThread(int(default_port_source), -1, default_host_source)
        thread.start()

# Detecs when a client disconnects
@socketio.on('disconnect', namespace='/test')
def test_disconnect():
    print('Client disconnected...')
    

# RESTFUL API for index
class Index(Resource):
    def __init__(self):
        pass
    def get(self):
        headers = {'Content-Type': 'text/html'}
        # allows the visualization of index if session logged in
        if not session.get('logged_in'):
            return make_response(render_template('login.html'), 200, headers)
        else:
            return make_response(render_template('index.html'), 200, headers)

# RESTFUL API for login
class Login(Resource):
    def __init__(self):
        pass
    def post(self):
        headers = {'Content-Type': 'text/html'}
        # Detects if the username/password are correct and starts the session
        if request.form['password'] == 'password' and request.form['username'] == 'admin':
            session['logged_in'] = True
            return make_response(render_template('index.html'), 200, headers)
        else:    
            flash("Try again...")
            return make_response(render_template('login.html'), 200, headers)

# RESTFUL API for logout
class Logout(Resource):
    def __init__(self):
        pass
    def get(self):
        session['logged_in'] = False
        return redirect(url_for('index'))

# Adds all the restful apis resources
api.add_resource(Index, '/')
api.add_resource(Login,'/login')
api.add_resource(Logout,'/logout')

if __name__ == '__main__':
    # Default host 
    default_host="127.0.0.1"
    # Default port
    default_port="5000"
    # Default host stream
    default_host_source="localhost"
    # Default port stream
    default_port_source="8888"
    # Parser of the options
    parser = optparse.OptionParser()
    parser.add_option("-H", "--host",
        help="Hostname of the Flask app " + \
            "[default %s]" % default_host,
        default=default_host)
    parser.add_option("-P", "--port",
        help="Port for the Flask app " + \
            "[default %s]" % default_port,
        default=default_port)
    
    parser.add_option("-O", "--source_host",
        help="host for the source generator " + \
            "[default %s]" % default_host_source,
        default=default_host_source)

    parser.add_option("-S", "--source_port",
        help="Port for the source generator " + \
            "[default %s]" % default_port_source,
        default=default_port_source)

    parser.add_option("-d", "--debug",
        action="store_true", dest="debug",
        help=optparse.SUPPRESS_HELP)
    options, _ = parser.parse_args()

    socketio.run(app,
        debug=options.debug,
        host=options.host,
        port=int(options.port))
